compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2646,37 +2644,241 @@ public Boolean visit(BParameterizedType t, BType s) {
 
     };
 
-    private boolean checkUnionHasSameFiniteType(LinkedHashSet<BType> memberTypes, BType baseType) {
-        for (BType type : memberTypes) {
-            if (type.tag != TypeTags.FINITE) {
+    private class BOrderedTypeVisitor implements BTypeVisitor<BType, Boolean> {
+
+        Set<TypePair> unresolvedTypes;
+
+        BOrderedTypeVisitor(Set<TypePair> unresolvedTypes) {
+            this.unresolvedTypes = unresolvedTypes;
+        }
+
+        @Override
+        public Boolean visit(BType target, BType source) {
+            if (isSimpleBasicType(source.tag) && isSimpleBasicType(target.tag)) {
+                return (source == target) || isIntOrStringType(source.tag, target.tag);
+            }
+            if (source.tag == TypeTags.FINITE) {
+                return checkValueSpaceHasSameType(((BFiniteType) source), target);
+            }
+            return false;
+        }
+
+        @Override
+        public Boolean visit(BArrayType target, BType source) {
+            if (target.tag != TypeTags.ARRAY || source.tag != TypeTags.ARRAY) {
                 return false;
             }
-            boolean isValueSpaceSameType = false;
-            for (BLangExpression expr : ((BFiniteType) type).getValueSpace()) {
-                isValueSpaceSameType = isSameType(expr.type, baseType);
-                if (!isValueSpaceSameType) {
-                    break;
+
+            BArrayType rhsArrayType = (BArrayType) source;
+            boolean hasSameTypeElements = isSameOrderedType(target.eType, rhsArrayType.eType, unresolvedTypes);
```

Comment:
The variable name doesn't quite reflect the actual check, right? Shall we rename it?

```suggestion
            boolean hasSameOrderedTypeElements = isSameOrderedType(target.eType, rhsArrayType.eType, unresolvedTypes);
```